src/blackdoor/cqbe/node/Updater.java

```diff
@@ -1,225 +1,244 @@
 package blackdoor.cqbe.node;
 
 import java.io.IOException;
-import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
 
 import blackdoor.cqbe.addressing.Address;
-import blackdoor.cqbe.addressing.Address.AddressComparator;
 import blackdoor.cqbe.addressing.AddressTable;
+import blackdoor.cqbe.addressing.CASFileAddress;
 import blackdoor.cqbe.addressing.L3Address;
 import blackdoor.cqbe.output_logic.Router;
 import blackdoor.cqbe.rpc.RPCException;
-import blackdoor.cqbe.storage.StorageController;
+import blackdoor.cqbe.rpc.RPCException.JSONRPCError;
 import blackdoor.util.DBP;
 
-
-/**
- * Responsibility - Continually update the address table of a node by pinging current AT members, removing non-responsive nodes,
- * 					and finding new neighbors to populate the address table.
- * @author Cyril Van Dyke
- * @version 1.1
- * 
- */
-
-public class Updater implements Runnable{
-	private Router r;
-	volatile Boolean timer;
-	private HashSet<Address> firstStrike = new HashSet<Address>();
-	private HashSet<Address> secondStrike = new HashSet<Address>();
+public class Updater implements Runnable {
 	
-	public Updater() {
-
+	/**
+	 * in seconds
+	 */
+	public static final long updateInterval = 5;
+	
+	public static final int PARALLELISM = 8;
+	
+	private Thread updaterThread;
+	private volatile boolean running;
+	private Map<L3Address, Integer> strikeList;
+	
+	public Updater(){
+		running = true;
+		strikeList = new ConcurrentHashMap<L3Address, Integer>();
 	}
-
 	
 
 	@Override
 	public void run() {
-		try {
-			r = new Router(Node.getAddressTable());
-			timer = true;
-			Node.getAddressTable().addAll(r.iterativeLookup(Node.getOverlayAddress()).values());
-			updateTimer(); 
-		} catch (Exception e) {
-			// TODO Auto-generated catch block
-			e.printStackTrace();
+		if(updaterThread == null){
+			updaterThread = Thread.currentThread();
 		}
-	}
-
-	
-	/**
-	 * Timer function that tells Updater when to check for updates
-	 * When updaterTimer fires, update() is run.
-	 * @throws UnknownHostException 
-	 */
-	private void updateTimer() throws UnknownHostException{
-		update();
-		try{
-			while (timer){
-				Thread.sleep(6000);
-				update();
+		
+		while(running){
+			try {
+				schedule();
+			} catch (InterruptedException e) {
+				DBP.printwarningln("Updater thread interrupted!");
+				run();
```

Comment:
The intention is to catch the interrupts not generated by stop(). That way the interrupt is caught, but running is still true, and the updater can continue.